test/jasmine/tests/histogram_test.js

```diff
@@ -664,3 +666,207 @@ describe('Test histogram', function() {
+        // 28-day gap STILL gets days - in principle this might happen with data
+        // that's actually monthly, if the bins edges fall on different months
+        // (ie not full years) but that's a very weird edge case so I'll ignore it!
+        _test(14 * day, 14 * day, [jan17 - 14 * day, jan18 - 14 * day, jan19 - 14 * day], 'gregorian',
+            [jan17 - 14 * day, jan18 - 15 * day, jan18 - 14 * day, jan19 - 15 * day]);
+        // 31-day gap - months
+        _test(15 * day, 16 * day, [jan17 - 15 * day, jan18 - 15 * day, jan19 - 15 * day], 'gregorian',
+            [jan17, Date.UTC(2017, 11, 1), jan18, Date.UTC(2018, 11, 1)]);
+        // 365-day gap - years have enough buffer to handle leap vs nonleap years
+        _test(182 * day, 183 * day, [jan17 - 182 * day, j2117 - 182 * day, j2217 - 182 * day], 'gregorian',
+            [jan17, j2116, j2117, j2216]);
+        // no bigger rounding than years
+        _test(20 * 365 * day, 20 * 365 * day, [jan17, j2117, j2217], 'gregorian',
+            [jan17, j2116, j2117, j2216]);
+    });
+
+    it('rounds (mostly) correctly when using world calendars', function() {
+        var cn = 'chinese';
+        var cn8 = Lib.dateTime2ms('1995-08-01', cn);
+        var cn8i = Lib.dateTime2ms('1995-08i-01', cn);
+        var cn9 = Lib.dateTime2ms('1995-09-01', cn);
+
+        var cn1_00 = Lib.dateTime2ms('2000-01-01', cn);
+        var cn1_01 = Lib.dateTime2ms('2001-01-01', cn);
+        var cn1_02 = Lib.dateTime2ms('2002-01-01', cn);
+        var cn1_10 = Lib.dateTime2ms('2010-01-01', cn);
+        var cn1_20 = Lib.dateTime2ms('2020-01-01', cn);
+
+        _test(100, 2000, [cn8i, cn8i + 10000, cn8i + 20000], cn,
+            [cn8i, cn8i + 9000, cn8i + 10000, cn8i + 19000]);
+        _test(500, 2500, [cn8i - 500, cn8i + 9500, cn8i + 19500], cn,
+            [cn8i, cn8i + 9000, cn8i + 10000, cn8i + 19000]);
+
+        _test(0, day, [cn8, cn8i, cn9], cn,
+            [cn8, cn8i - day, cn8i, cn9 - day]);
+        _test(12 * hr, 12 * hr, [cn8 - 12 * hr, cn8i - 12 * hr, cn9 - 12 * hr], cn,
+            [cn8, cn8i - day, cn8i, cn9 - day]);
+
+        _test(0, 28 * day, [cn1_00, cn1_01, cn1_02], cn,
+            [cn1_00, Lib.dateTime2ms('2000-12-01', cn), cn1_01, Lib.dateTime2ms('2001-12-01', cn)]);
+        _test(14 * day, 14 * day, [cn1_00 - 14 * day, cn1_01 - 14 * day, cn1_02 - 14 * day], cn,
+            [cn1_00, Lib.dateTime2ms('2000-12-01', cn), cn1_01, Lib.dateTime2ms('2001-12-01', cn)]);
+
+        _test(0, 353 * day, [cn1_00, cn1_10, cn1_20], cn,
+            [cn1_00, Lib.dateTime2ms('2009-01-01', cn), cn1_10, Lib.dateTime2ms('2019-01-01', cn)]);
+        // occasionally we give extra precision for world dates (month when it should be year
+        // or day when it should be month). That's better than doing the opposite... not going
+        // to fix now, too many edge cases, better not to complicate the logic for them all.
```

Comment:
The algorithm I came up with for figuring out how many digits to show is basically:
- find the minimum distance from data points to bin edges from either side - this is just two numbers for the whole distribution
- for two consecutive bins, find the largest digit that changes in both the left and right gaps (with a little shifting to account for rounding errors)

This algorithm is a little complicated to state, but it's quick to calculate and it's robust as far as I can see, on numeric, category, and gregorian date axes (though even in that case there is a weird [edge case](https://github.com/plotly/plotly.js/pull/2113/commits/6cb4e7130e4a5a1de7e5819fc6eb7c487c7997a4#diff-8c89958f1286cac36ab2325921671724R819) that confuses it). However, the larger variability of month and year lengths in world calendars causes it to sometimes think a digit doesn't change when in the actual data it does, and increasing the tolerance to catch this breaks some much more common cases.

I suspect we could fix this by looking at *all* bin edges independent of each other, but that's far more computationally intensive, and has some other potential weird pitfalls. I don't think providing a bit of extra precision in the label for a few world calendar cases justifies all that extra complexity.

Reply:
The issue with dates is that the actual time increments - the "digits" we're looking to see if they have changed - have different sizes in terms of milliseconds. The *bin sizes* being different should not be a problem, though people could do things to thwart the algorithm, like having the first two bin edges on fairly round values but some subsequent edges at not-so-round values. The idea that two edges are enough to determine the biggest digit that always changes comes from the observation that if *one* bin is extra-round, its neighbor will not be. So to do this robustly with nonuniform bins, we could either:
- Do this same algorithm on *all* bin edges. That's different from looking at all bin edges independently as I suggested above for world calendars, though that would solve this problem too - what I'm suggesting here is to calculate just one left-edge-gap and one right-edge-gap as we do now, but then test those gaps against all bin edges instead of the first two.
- As a last resort, make a new attribute for folks to set their own rounding.

One common case mentioned in #360 is equally-sized bins on a log axis. There, the whole assumption of a single digit to round to is wrong, and we'll have to do something completely different or just give up on the idea of disambiguating the edges, and let the regular axis-formatting-for-hover system handle it unchanged.